crates/circuit/src/parameterexpression.rs

```diff
@@ -0,0 +1,779 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2023, 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+// symbol_exprpy.rs
+// Python interface of symbolic expression
+use crate::symbol_expr::{SymbolExpr, Value, SYMEXPR_EPSILON};
+use crate::symbol_parser::parse_expression;
+
+use hashbrown::{HashMap, HashSet};
+use num_complex::Complex64;
+
+use std::collections::hash_map::DefaultHasher;
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Neg, Sub, SubAssign};
+
+use pyo3::prelude::*;
+use pyo3::IntoPyObjectExt;
+
+// Python interface to SymbolExpr
+#[pyclass(sequence, module = "qiskit._accelerate.circuit")]
+#[derive(Clone, Debug)]
+pub struct ParameterExpression {
+    expr: SymbolExpr,
+}
+
+#[pymethods]
+impl ParameterExpression {
+    /// parse expression from string
+    #[new]
+    #[pyo3(signature = (in_expr=None))]
+    pub fn new(in_expr: Option<String>) -> PyResult<Self> {
+        match in_expr {
+            Some(e) => Ok(ParameterExpression {
+                expr: parse_expression(&e),
+            }),
+            None => Ok(ParameterExpression {
+                expr: SymbolExpr::Value(Value::Int(0)),
+            }),
+        }
+    }
+
+    /// create new expression as a symbol
+    #[allow(non_snake_case)]
+    #[staticmethod]
+    pub fn Symbol(name: String) -> Self {
+        // check if expr contains replacements for sympy
+        let name = name
+            .replace("__begin_sympy_replace__", "$\\")
+            .replace("__end_sympy_replace__", "$");
+
+        ParameterExpression {
+            expr: SymbolExpr::Symbol(Box::new(name)),
+        }
+    }
+
+    /// create new expression as a value
+    #[allow(non_snake_case)]
+    #[staticmethod]
+    pub fn Value(py: Python, value: PyObject) -> PyResult<Self> {
+        match Self::_extract_value(py, value) {
+            Some(v) => Ok(v),
+            None => Err(pyo3::exceptions::PyRuntimeError::new_err(
+                "Unsupported data type to initialize SymbolExpr as a value",
+            )),
+        }
+    }
+
+    #[inline]
+    #[staticmethod]
+    fn _extract_value(py: Python, value: PyObject) -> Option<Self> {
```

Comment:
It's probably better here to just accept `value` as a `Bound`, e.g.:

```suggestion
    fn _extract_value(value: &Bound<PyAny>) -> Option<Self> {
```

This way, you don't need to pass `py` to all of the `extract` calls.